GeoffreyBooth-proposal.md

```diff
@@ -0,0 +1,78 @@
+
+### Modernizing Output
+
+Lastly, CoffeeScript’s output should be modernized as much as possible. Fat arrows should output as fat arrows, destructuring should output as destructuring, etc. But I can’t emphasize enough that this should be the **last priority**—it makes our source code cleaner, and debugging eventually easier when ES2015 features are supported natively in runtimes, but otherwise ES2015 output has little benefit for developers. That said, *new* features, especially ES2015+ ones, should definitely be output as ES2015+ code.
+
+## A New Compiler
+
+The current CoffeeScript 1.x compiler is extremely well commented, but it really needs those comments. Understanding it will bend your brain into knots. The original goal of [CoffeeScript Redux](https://github.com/michaelficarra/CoffeeScriptRedux) was to refactor CoffeeScript’s codebase, not to add new features.
+
+Don’t underestimate how hard it is to write your own language compiler. If it were easy, CoffeeScript Redux would have been finished; as it stands, that project was essentially abandoned a year ago.
+
+Also don’t underestimate how hard it is to add new features to the CoffeeScript 1.x compiler. If that were easy, someone else would’ve added modules and classes to it by now.
+
+The CoffeeScript community suffers from too many people with a lot of passion but little time to invest in development. For that reason, I consider a new-from-scratch compiler to be off the table. We’ll just end up with another Redux.
+
+So we have only two options for starting points for our new compiler:
+
+* The current CoffeeScript 1.x compiler.
+	* Pros: already supports all existing CoffeeScript code.
+	* Cons: complicated, difficult-to-understand codebase.
+* The CoffeeScript Redux compiler, probably from [this PR](https://github.com/michaelficarra/CoffeeScriptRedux/pull/344).
+	* Pros: better codebase, and the PR already supports many ES2015 features.
+	* Cons: we inherit Redux’s issues and unfinished work, including any incompatibilities with existing CoffeeScript code.
+
+I don’t have a recommendation as to one or the other. I invite contributors to both projects to chime in.
+
+## Workflow
+
+As to where we should put our new compiler and develop it, I have two ideas: 
+
+**Inside the current main [coffee-script repo](https://github.com/jashkenas/coffeescript).** Currently the repo is structured such that the code in `src` is compiled and output into `lib/coffee-script`. If we want to avoid merge conflicts with any pending PRs against the repo, we should leave those folders as they are.
+
+But we could add new folders, for example `source` and `lib/coffeescript`, that contain the new compiler source code and compiled output. A flag like `output: 'modern'` would cause the new compiler (i.e. the code in `lib/coffeescript`) to be used instead of the old one. Both compilers, modern and legacy, would live on in the same repo. Bugfixes and improvements could be submitted against either.
```

Comment:
I want to avoid both the name `es6`, which is obsolete, and the notion that we’re targeting a specific version of JavaScript/ECMAScript. The output of the new compiler will be whatever the latest ECMAScript standard is. If it’s built now, it will output ES2015. When ES2016 is finalized and we add support for any new features that ES2016 adds that we decide to adopt, then the compiler will output ES2016. It will _not,_ however, have options for outputting ES5 or ES2015 or ES2016—developers’ only choices will be the old compiler, which will basically be ES5 plus generators and modules and maybe classes; or the new compiler, which would output whatever the latest JavaScript is that Babel can take as input.

There will be an overlap period where there’s some development on _both_ the old ES5 compiler and the new ES-latest compiler. For that reason, we can’t just keep the `coffee-script` repo and bump to version 2 and the new compiler is 2.x.x. We need a way to opt into it, either with a flag on the same `coffee-script` repo/NPM module; or a new repo/module under the `coffeescript` name.